### ISSUE TITLE
pwndbg.typeinfo.update needs to be invoked on new_objfile

### ISSUE THREAD

**Issue by zachriggle** [pwndbg/pwndbg#766]
When debugging a binary that does an `execve`, when using `gdbserver`, there is some stale state that gets left behind.  I expect some data is being over-cached or some events are not firing.

Specifically, this occurs when the exec'ed binary is of a different architecture.

You should be able to reproduce this with a 32-bit binary that invokes e.g. `system('id')`.

The attached binary simply invokes `execve("/bin/sh", 0, 0)`: [i386-linux-binsh.gz](https://github.com/pwndbg/pwndbg/files/4694602/i386-linux-binsh.gz).  Since `/bin/sh` is a 64-bit binary, things go a bit awry.

This only occurs when the original process is launched under `gdbserver`.  The process of launching the binary under gdbserver can be automated with Pwntools:

```py
gdb.debug('i386-linux-binsh').interactive()
```

## Initial state

```
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
─────────────────────────────[ REGISTERS ]─────────────────────────────
 EAX  0x0
 EBX  0x0
 ECX  0x0
 EDX  0x0
 EDI  0x0
 ESI  0x0
 EBP  0x0
 ESP  0xffd3dd30 ◂— 0x1
 EIP  0x10000 (_start) ◂— push   0x68 /* 0x2f68686a */
──────────────────────────────[ DISASM ]───────────────────────────────
 ► 0x10000 <_start>       push   0x68
   0x10002 <_start+2>     push   0x732f2f2f
   0x10007 <_start+7>     push   0x6e69622f
   0x1000c <_start+12>    mov    ebx, esp
   0x1000e <_start+14>    push   0x1010101
   0x10013 <_start+19>    xor    dword ptr [esp], 0x1016972
   0x1001a <_start+26>    xor    ecx, ecx
   0x1001c <_start+28>    push   ecx
   0x1001d <_start+29>    push   4
   0x1001f <_start+31>    pop    ecx
   0x10020 <_start+32>    add    ecx, esp
───────────────────────────────[ STACK ]───────────────────────────────
00:0000│ esp  0xffd3dd30 ◂— 0x1
01:0004│      0xffd3dd34 —▸ 0xffd3ede6 ◂— 0x2d667463 ('ctf-')
02:0008│      0xffd3dd38 ◂— 0x0
03:000c│      0xffd3dd3c —▸ 0xffd3ee05 ◂— 0x3d202020 ('   =')
04:0010│      0xffd3dd40 —▸ 0xffd3ee26 ◂— 0x752f3d5f ('_=/u')
05:0014│      0xffd3dd44 —▸ 0xffd3ee3f ◂— 0x4c524550 ('PERL')
06:0018│      0xffd3dd48 —▸ 0xffd3ee6c ◂— 0x5353454c ('LESS')
07:001c│      0xffd3dd4c —▸ 0xffd3ee8e ◂— 0x5353454c ('LESS')
─────────────────────────────[ BACKTRACE ]─────────────────────────────
 ► f 0    10000 _start
pwndbg>
```

## After `continue`

Here we can see that the code mappings are kind of broken.  However, `vmmap` has correct information.  Note that the registers shown are still 32-bit.

```
pwndbg> c
Continuing.

Program received signal SIGTRAP, Trace/breakpoint trap.
ERROR: Could not find ELF base!
0x1d66a090 in ?? ()
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
─────────────────────────────[ REGISTERS ]─────────────────────────────
 EAX  0x0
 EBX  0x0
 ECX  0x0
 EDX  0x0
 EDI  0x0
 ESI  0x0
 EBP  0x0
 ESP  0xd4be1680
 EIP  0x1d66a090
──────────────────────────────[ DISASM ]───────────────────────────────
Invalid address 0x1d66a090










───────────────────────────────[ STACK ]───────────────────────────────
<Could not read memory at 0xd4be1680>
─────────────────────────────[ BACKTRACE ]─────────────────────────────
 ► f 0 1d66a090
pwndbg> vmmap
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
0x561691661000 0x56169167d000 r-xp    1c000 0      /bin/dash
0x56169187c000 0x56169187f000 rw-p     3000 1b000  /bin/dash
0x56169187f000 0x561691881000 rw-p     2000 0
0x7f111d669000 0x7f111d690000 r-xp    27000 0      /lib/x86_64-linux-gnu/ld-2.27.so
0x7f111d890000 0x7f111d892000 rw-p     2000 27000  /lib/x86_64-linux-gnu/ld-2.27.so
0x7f111d892000 0x7f111d893000 rw-p     1000 0
0x7ffcd4bc1000 0x7ffcd4be2000 rw-p    21000 0      [stack]
0x7ffcd4be7000 0x7ffcd4bea000 r--p     3000 0      [vvar]
0x7ffcd4bea000 0x7ffcd4bec000 r-xp     2000 0      [vdso]
0xffffffffff600000 0xffffffffff601000 r-xp     1000 0      [vsyscall]
```

**Comment by disconnect3d** [comment]
Interesting, can u see this with:
```
pi import pwndbg
pi pwndbg.memoize.memoize.caching = False
```
?

**Comment by zachriggle** [comment]
The same behavior is observed with caching disabled.

The same behavior is actually observed with Pwndbg completely disabled ☹️ 

```
Reading symbols from /tmp/pwntools/pwn-asm-ubO8i4/step3...(no debugging symbols found)...done.
0x10000000 in _start ()
(gdb) c
Continuing.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x4ba88090 in ?? ()
(gdb)
```

**Comment by zachriggle** [comment]
I verified that this issue still affects GDB 9.2 ☹️

**Comment by zachriggle** [comment]
Following up, even with GDB 9.2 I was still using an old (7.10) gdbserver that was in my ~/bin/.

Updating gdbserver to also be 9.2 makes everything work again (mostly).

**Comment by zachriggle** [comment]
Looks like this IS pwndbg-related, at least partly.  We are not updating register widths after an exec.  I think there's a missing event or something like that?

**TLDR: We need to add `@pwndbg.events.new_objfile` to `pwndbg.typeinfo.update`**

## Test Script

```test.c
// gcc -m32 -static -g test.c
#include <unistd.h>
int main() {
    execve("/bin/sh", 0, 0);
}
```

## Without Pwndbg

Once `/bin/sh` is executed and you see the prompt, hit Ctrl+C to get back to the debugger.

```
$ gdb ./a.out
Reading symbols from ./a.out...
(gdb) b main
Breakpoint 1 at 0x80488be: file test.c, line 3.
(gdb) r
Starting program: /home/pwntools/a.out

Breakpoint 1, main () at test.c:3
3	    execve("/bin/sh", 0, 0);
(gdb) c
Continuing.
process 937 is executing new program: /bin/dash
Error in re-setting breakpoint 1: Function "main" not defined.
warning: Could not load shared library symbols for linux-vdso.so.1.
Do you need "set solib-search-path" or "set sysroot"?
$ ^C
Program received signal SIGINT, Interrupt.
0x00007f0b53518081 in read () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) i r rax
rax            0xfffffffffffffe00  -512
(gdb) show arch
The target architecture is set automatically (currently i386:x86-64)
```

## With Pwndbg

Note that after the exec and Ctrl+C, the register *names* are correct, but all of the widths are wrong.

```
$ gdb ./a.out
pwndbg: loaded 174 commands. Type pwndbg [filter] for a list.
pwndbg: created $rebase, $ida gdb functions (can be used with print/break)
Reading symbols from ./a.out...

pwndbg> b main
Breakpoint 1 at 0x80488be: file test.c, line 3.

pwndbg> r
Starting program: /home/pwntools/a.out

Breakpoint 1, main () at test.c:3
3	    execve("/bin/sh", 0, 0);
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
────────────────────────────────────────[ REGISTERS ]────────────────────────────────────────
 EAX  0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
 EBX  0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
 ECX  0xffb4fa20 ◂— 0x1
 EDX  0xffb4fa74 —▸ 0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
 EDI  0x80481a8 (_init) ◂— push   ebx
 ESI  0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
 EBP  0xffb4fa08 ◂— 0x0
 ESP  0xffb4fa00 —▸ 0xffb4fa20 ◂— 0x1
 EIP  0x80488be (main+25) ◂— sub    esp, 4
─────────────────────────────────────────[ DISASM ]──────────────────────────────────────────
 ► 0x80488be <main+25>    sub    esp, 4
   0x80488c1 <main+28>    push   0
   0x80488c3 <main+30>    push   0
   0x80488c5 <main+32>    lea    edx, [eax - 0x2d238]
   0x80488cb <main+38>    push   edx
   0x80488cc <main+39>    mov    ebx, eax
   0x80488ce <main+41>    call   execve <0x806c020>

   0x80488d3 <main+46>    add    esp, 0x10
   0x80488d6 <main+49>    mov    eax, 0
   0x80488db <main+54>    lea    esp, [ebp - 8]
   0x80488de <main+57>    pop    ecx
──────────────────────────────────────[ SOURCE (CODE) ]──────────────────────────────────────
In file: /home/pwntools/test.c
   1 #include <unistd.h>
   2 int main() {
 ► 3     execve("/bin/sh", 0, 0);
   4 }
──────────────────────────────────────────[ STACK ]──────────────────────────────────────────
00:0000│ esp  0xffb4fa00 —▸ 0xffb4fa20 ◂— 0x1
01:0004│      0xffb4fa04 —▸ 0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
02:0008│ ebp  0xffb4fa08 ◂— 0x0
03:000c│      0xffb4fa0c —▸ 0x8048f0f (__libc_start_main+943) ◂— add    esp, 0x10
04:0010│      0xffb4fa10 —▸ 0x806ea7f (_dl_debug_initialize+15) ◂— add    ecx, 0x6a581
05:0014│      0xffb4fa14 —▸ 0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
... ↓
07:001c│      0xffb4fa1c —▸ 0x8048f0f (__libc_start_main+943) ◂— add    esp, 0x10
────────────────────────────────────────[ BACKTRACE ]────────────────────────────────────────
 ► f 0  80488be main+25
   f 1  8048f0f __libc_start_main+943
Breakpoint main

pwndbg> c
Continuing.
process 1014 is executing new program: /bin/dash
Error in re-setting breakpoint 1: Function "main" not defined.
warning: Could not load shared library symbols for linux-vdso.so.1.
Do you need "set solib-search-path" or "set sysroot"?
$ ^C
Program received signal SIGINT, Interrupt.
0x00007ff9c7eda081 in __GI___libc_read (fd=0x0, buf=0x55e4efdc4a60 <basebuf>, nbytes=0x2000) at ../sysdeps/unix/sysv/linux/read.c:27
27	../sysdeps/unix/sysv/linux/read.c: No such file or directory.
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
────────────────────────────────────────[ REGISTERS ]────────────────────────────────────────
 RAX  0xfffffe00
 RBX  0x0
 RCX  0xc7eda081
 RDX  0x2000
 RDI  0x0
 RSI  0xefdc4a60
 R8   0xf0f0638b
 R9   0x0
 R10  0xf0f06010
 R11  0x246
 R12  0xefbbd109
 R13  0x0
 R14  0x1
 R15  0x0
 RBP  0xefdc4a60
 RSP  0x7ba734f8
 RIP  0xc7eda081
─────────────────────────────────────────[ DISASM ]──────────────────────────────────────────
Invalid address 0xc7eda081
──────────────────────────────────────────[ STACK ]──────────────────────────────────────────
<Could not read memory at 0x7ba734f8>
────────────────────────────────────────[ BACKTRACE ]────────────────────────────────────────
 ► f 0 c7eda081 read+17
   f 1 efbb217d
   f 2 efbb6a15
   f 3 efbb6c15
   f 4 efbb89c7
   f 5 efbb45ef
   f 6 efbab9de
   f 7 c7debb97 __libc_start_main+231
Program received signal SIGINT
```

Note that the architecture is correct and GDB is returning correct values.

```
pwndbg> show arch
The target architecture is set automatically (currently i386:x86-64)

pwndbg> i r eax rax
eax            0xfffffe00          0xfffffe00
rax            0xfffffffffffffe00  0xfffffffffffffe00
```

However, internally Pwndbg has the wrong register widths.

```
pwndbg> pi
>>> pwndbg.arch.current
'x86-64'
>>> hex(pwndbg.regs.sp)
'0x7ba734f8'
>>> hex(pwndbg.regs.esp)
'0x7ba734f8'
>>> hex(pwndbg.regs.rsp)
'0x7ba734f8'
```

This seems odd, because GDB has the correct values

```
>>> hex(int(gdb.newest_frame().read_register('esp')))
'0x7ba734f8'
>>> hex(int(gdb.newest_frame().read_register('rsp')))
'0x7ffe7ba734f8'
```

This occurs even with `pwndbg.memoize.memoize.caching = False`

It looks like we are missing a `pwndbg.typeinfo.update()` call.  Calling it manually makes everything work.

**Comment by zachriggle** [comment]
This was extra hard to debug because there IS (was) a bug in gdbserver that breaks architecture detection across execs.  It seems to be fixed in v9.2, but is definitely broken in v7.10.

Thanks for the patience reading this all, I renamed the issue and put in a TLDR.

**Comment by disconnect3d** [comment]
> TLDR: We need to add @pwndbg.events.new_objfile to pwndbg.typeinfo.update

@zachriggle Can it be that you tested it with old pwndbg? The `pwndbg.typeinfo.update` is fired upon `new_objfile` event, as in https://github.com/pwndbg/pwndbg/blob/dev/pwndbg/typeinfo.py#L46

Also, the PoC with 32-bit binary that calls execve on /bin/sh which is 64-bit works properly for me. This being said, I am closing this unless more info is shown.

Log:
```
dc@jhtc:~$ cat a.c
#include <unistd.h>
int main() {
    execve("/bin/sh", 0, 0);
}

dc@jhtc:~$ gcc -m32 -static -g a.c
a.c: In function 'main':
a.c:3:5: warning: null argument where non-null required (argument 2) [-Wnonnull]
     execve("/bin/sh", 0, 0);
     ^~~~~~
dc@jhtc:~$ file a.out
a.out: ELF 32-bit LSB executable, Intel 80386, version 1 (GNU/Linux), statically linked, for GNU/Linux 3.2.0, BuildID[sha1]=65703bc16b57652694477f564f1ce51e381ddf7f, with debug_info, not stripped
dc@jhtc:~$ cat ~/.gdbinit
source /opt/pwndbg/gdbinit.py
dc@jhtc:~$ gdb ./a.out
GNU gdb (Ubuntu 8.1-0ubuntu3.2) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
******
Your encoding (ANSI_X3.4-1968) is different than UTF-8. pwndbg might not work properly.
You might try launching gdb with:
    LC_ALL=en_US.UTF-8 PYTHONIOENCODING=UTF-8 gdb
Make sure that en_US.UTF-8 is activated in /etc/locale.gen and you called locale-gen
******
pwndbg: loaded 193 commands. Type pwndbg [filter] for a list.
pwndbg: created $rebase, $ida gdb functions (can be used with print/break)
Reading symbols from ./a.out...done.
pwndbg> b main
Breakpoint 1 at 0x80488be: file a.c, line 3.
pwndbg> r
Starting program: /home/dc/a.out

Breakpoint 1, main () at a.c:3
3	    execve("/bin/sh", 0, 0);
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
────────────────────────────────────────────[ REGISTERS ]────────────────────────────────────────────
 EAX  0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
 EBX  0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
 ECX  0xffffd2e0 ◂— 0x1
 EDX  0xffffd334 —▸ 0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
 EDI  0x80481a8 (_init) ◂— push   ebx
 ESI  0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
 EBP  0xffffd2c8 ◂— 0x0
 ESP  0xffffd2c0 —▸ 0xffffd2e0 ◂— 0x1
 EIP  0x80488be (main+25) ◂— sub    esp, 4
─────────────────────────────────────────────[ DISASM ]──────────────────────────────────────────────
 ► 0x80488be <main+25>    sub    esp, 4
   0x80488c1 <main+28>    push   0
   0x80488c3 <main+30>    push   0
   0x80488c5 <main+32>    lea    edx, [eax - 0x2d258]
   0x80488cb <main+38>    push   edx
   0x80488cc <main+39>    mov    ebx, eax
   0x80488ce <main+41>    call   execve <0x806c010>

   0x80488d3 <main+46>    add    esp, 0x10
   0x80488d6 <main+49>    mov    eax, 0
   0x80488db <main+54>    lea    esp, [ebp - 8]
   0x80488de <main+57>    pop    ecx
──────────────────────────────────────────[ SOURCE (CODE) ]──────────────────────────────────────────
In file: /home/dc/a.c
   1 #include <unistd.h>
   2 int main() {
 ► 3     execve("/bin/sh", 0, 0);
   4 }
──────────────────────────────────────────────[ STACK ]──────────────────────────────────────────────
00:0000│ esp  0xffffd2c0 —▸ 0xffffd2e0 ◂— 0x1
01:0004│      0xffffd2c4 —▸ 0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
02:0008│ ebp  0xffffd2c8 ◂— 0x0
03:000c│      0xffffd2cc —▸ 0x8048f0f (__libc_start_main+943) ◂— add    esp, 0x10
04:0010│      0xffffd2d0 —▸ 0x806ea6f (_dl_debug_initialize+15) ◂— add    ecx, 0x6a591
05:0014│      0xffffd2d4 —▸ 0x80d9000 (_GLOBAL_OFFSET_TABLE_) ◂— 0x0
... ↓
07:001c│      0xffffd2dc —▸ 0x8048f0f (__libc_start_main+943) ◂— add    esp, 0x10
────────────────────────────────────────────[ BACKTRACE ]────────────────────────────────────────────
 ► f 0  80488be main+25
   f 1  8048f0f __libc_start_main+943
─────────────────────────────────────────────────────────────────────────────────────────────────────
pwndbg> # regs are 32-bit whicch is correct
pwndbg> c
Continuing.
process 12828 is executing new program: /bin/dash
Error in re-setting breakpoint 1: Function "main" not defined.
$ ^C
Program received signal SIGINT, Interrupt.
0x00007ffff7af4081 in __GI___libc_read (fd=0, buf=0x555555771a60 <basebuf>, nbytes=8192) at ../sysdeps/unix/sysv/linux/read.c:27
27	../sysdeps/unix/sysv/linux/read.c: No such file or directory.
ERROR: Could not find ELF base!
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
────────────────────────────────────────────[ REGISTERS ]────────────────────────────────────────────
 RAX  0xfffffffffffffe00
 RBX  0x0
 RCX  0x7ffff7af4081 (read+17) ◂— cmp    rax, -0x1000 /* 'H=' */
 RDX  0x2000
 RDI  0x0
 RSI  0x555555771a60 (basebuf) ◂— 0x0
 R8   0x55555577438b ◂— 0x20240000000000
 R9   0x0
 R10  0x555555774010 ◂— 0x100000000
 R11  0x246
 R12  0x55555556a109 ◂— jae    0x55555556a173 /* 'sh: turning off NDELAY mode\n' */
 R13  0x0
 R14  0x1
 R15  0x0
 RBP  0x555555771a60 (basebuf) ◂— 0x0
 RSP  0x7fffffffebc8 —▸ 0x55555555f17d ◂— test   eax, eax
 RIP  0x7ffff7af4081 (read+17) ◂— cmp    rax, -0x1000 /* 'H=' */
─────────────────────────────────────────────[ DISASM ]──────────────────────────────────────────────
 ► 0x7ffff7af4081 <read+17>     cmp    rax, -0x1000
   0x7ffff7af4087 <read+23>     ja     read+112 <0x7ffff7af40e0>
    ↓
   0x7ffff7af40e0 <read+112>    mov    rdx, qword ptr [rip + 0x2dad81]
   0x7ffff7af40e7 <read+119>    neg    eax
   0x7ffff7af40e9 <read+121>    mov    dword ptr fs:[rdx], eax
   0x7ffff7af40ec <read+124>    mov    rax, -1
   0x7ffff7af40f3 <read+131>    ret

   0x7ffff7af40f4 <read+132>    mov    rdx, qword ptr [rip + 0x2dad6d]
   0x7ffff7af40fb <read+139>    neg    eax
   0x7ffff7af40fd <read+141>    mov    dword ptr fs:[rdx], eax
   0x7ffff7af4100 <read+144>    mov    rax, -1
──────────────────────────────────────────────[ STACK ]──────────────────────────────────────────────
00:0000│ rsp  0x7fffffffebc8 —▸ 0x55555555f17d ◂— test   eax, eax
01:0008│      0x7fffffffebd0 —▸ 0x7fffffffec30 —▸ 0x555555773aa0 (stackbase) ◂— 0x0
02:0010│      0x7fffffffebd8 ◂— 0x0
... ↓
04:0020│      0x7fffffffebe8 —▸ 0x555555563a15 ◂— mov    ebx, eax
05:0028│      0x7fffffffebf0 —▸ 0x7fffffffec30 —▸ 0x555555773aa0 (stackbase) ◂— 0x0
06:0030│      0x7fffffffebf8 —▸ 0x555555563c15 ◂— test   r12d, r12d
07:0038│      0x7fffffffec00 —▸ 0x7fffffffec30 —▸ 0x555555773aa0 (stackbase) ◂— 0x0
────────────────────────────────────────────[ BACKTRACE ]────────────────────────────────────────────
 ► f 0     7ffff7af4081 read+17
   f 1     55555555f17d
   f 2     555555563a15
   f 3     555555563c15
   f 4     5555555659c7
   f 5     5555555615ef
   f 6     5555555589de
   f 7     7ffff7a05b97 __libc_start_main+231
─────────────────────────────────────────────────────────────────────────────────────────────────────
pwndbg> # regs are 64--bit which is correct after execve
```
